### Bakta


```{r, echo=FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(DT)
library(knitr)
library(plotly)
```
```{r, echo=FALSE}
BAKTA_INPUT_FILE <- "#RESULTSENV#"
SAMPLE_DATA <- "#NAMEENV#"
BAKTA_VERSION <- "#VERSIONENV#"
PATH <- "#PATHENV#" 

BAKTA_RAW <- read_delim(BAKTA_INPUT_FILE, 
    delim = "\t", escape_double = FALSE,
    col_names = FALSE, comment = c("#"), trim_ws = TRUE)
BAKTA_RAW <- na.omit(BAKTA_RAW)
```

```{r}
#Sanity checks
#EXIT <- FALSE
#Check if empty file/structured in a not bakta way 
#if (nrow(BAKTA_RAW)<2){
#print("Source file seems empty, check if annotation file was created correctly")
#  EXIT <- TRUE
#}
```

```{r}
#Abort if fail sanity checks
#if (EXIT == TRUE){
#knit_exit()
#}
```

```{r}
# Parsing gff
BAKTA_PROC <- separate(BAKTA_RAW, X9, c(NA, "name"), remove = FALSE, "Name=")
BAKTA_PROC$name <- gsub(";.*","",BAKTA_PROC$name)
BAKTA_PROC <- BAKTA_PROC %>% filter(X3 != "gene")
```

```{r}
GEN_N <- BAKTA_PROC %>% nrow() %>% as.integer()
CONTIG_N <- length(unique(BAKTA_PROC$X1))
PIE <- BAKTA_PROC %>% filter(X3 != 'region') %>%  count( ,X3 ) %>% as.data.frame()
RRNA_N_GEN <- PIE %>% filter(X3 == 'rRNA') %>% select(n) %>% as.integer()
TRNA_N_GEN <- PIE %>% filter(X3 == 'tRNA') %>% select(n) %>% as.integer()
if (is.na(RRNA_N_GEN)){
  RRNA_N_GEN <- 0
}
if (is.na(TRNA_N_GEN)){
  TRNA_N_GEN <- 0
}
#rRNA count
ARNA <- BAKTA_PROC %>% filter(name == '5S ribosomal RNA') %>%  nrow()
BRNA <- BAKTA_PROC %>% filter(name == '16S ribosomal RNA') %>%  nrow()
CRNA <- BAKTA_PROC %>% filter(name == '23S ribosomal RNA') %>%  nrow()
```
Bakta is a tool for the rapid & standardized annotation of bacterial genomes & plasmids. 
[BAKTA Github](https://github.com/oschwengers/bakta)  

#### Data
Sample name: **`r SAMPLE_DATA`**

Bakta version used:  **`r BAKTA_VERSION`**

Results are stored in: **`r PATH`**

#### Overview
Bakta founded **`r GEN_N` features**  Features are located on **`r CONTIG_N` contigs**  (see Fig.2) and **`r TRNA_N_GEN` tRNA** sequences.

Bacterial species can contain up to 15 rRNA operons [(Klappenbach et al.)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC91988/#B28) each containing the three RNA genes: 5S, 16S and 23S in a close proximity. 

Bakta detected `r ARNA` times 5S, `r BRNA` times 16S and `r CRNA` times 23S rRNA genes (this would correspond to ~`r as.integer((ARNA+BRNA+CRNA)/3)` operons) <span style="color:red; font-weight: bold">`r if (((ARNA+BRNA+CRNA)/3)>15){
"Warning found more than 15 operons!"
}`</span><span style="color:red; font-weight: bold">`r if (all(sapply(list(ARNA,BRNA,CRNA), function(x) x != CRNA))){
"Warning some rRNA operons are missing one of the rRNA genes!"
}`</span>. You can check their close proximity via Table 3 (search "ribosomal RNA"). If you have more RNA operons than expected please check the genome quality or if contaminations from other organisms are present increasing the rRNA count.


Fig.1  Pie chart representing all the annotation features identified by Bakta
```{r}
#margins
M <- list(
          l = 50,
          r = 50,
          b = 50,
          t = 50,
          pad = 4
        )
COMP_FIG <- plot_ly(PIE, labels = ~X3, values = ~n, width = 450, height = 450) %>% 
    add_pie(hole = 0.6) %>% layout(showlegend = TRUE,
                                    margin = M,
                                    legend = list(orientation = 'h'),
                                  xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
                                  yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))
COMP_FIG
```

Clicking on the legend will exclude the feature from the composition and clicking in the right corner on camera allows to download currently displayed plot as a png file.  

#### Composition and contig length plots
Visualisation of features found on contigs compared with length of contigs.

Fig. 2 Composition by annotation features of each contig. Hovering over a bar will display more details about the element. The plot is zoomable by highlighting a region that you want to zoom into. Clicking at the top right corner to download the currently displayed plot as a png file.
```{r include = FALSE}
#### Data conversion 
CONT_LEN <- BAKTA_PROC %>% select(X1, X5) %>% group_by(X1) %>% summarise(
                                                                          cont_max = max(X5)
                                                                          )
CONT_COMP <- BAKTA_PROC %>% filter(X3 != "region") %>%select(X1, X3)
CONT_TALLY <- CONT_COMP %>% select(X1, X3) %>% count(X1,X3)
CONT_TALLY <-spread(CONT_TALLY, key = "X3", value = "n")
CONT_TALLY <- merge(CONT_LEN,CONT_TALLY, by="X1" )
CONT_TALLY$cont_max <- as.integer(CONT_TALLY$cont_max)
order(CONT_TALLY[,"cont_max"])
CONT_TALLY <- CONT_TALLY %>%
  select(X1,cont_max,CDS, everything())
CONT_TALLY <- CONT_TALLY %>% arrange(desc(cont_max))
if (!"CDS" %in% colnames(CONT_TALLY)) {
  CONT_TALLY$CDS <- 0
}
if (!"tRNA" %in% colnames(CONT_TALLY)) {
  CONT_TALLY$tRNA <- 0
}
if (!"rRNA" %in% colnames(CONT_TALLY)) {
  CONT_TALLY$rRNA <- 0
}
#### Fig dimensions
len <- nrow(CONT_LEN)
dim <- 0 
if(len > 10){
dim <- as.numeric((len - 10) *100) 
}
dim_high <- as.numeric(600+dim) 
```

```{r}
#### Plots 
#### Comp Plot
CONT_COMP_FIG <- plot_ly(CONT_TALLY, y = ~X1, x = ~CDS, type = 'bar', orientation = 'h',width = 950, height = dim_high,
                      text = CONT_TALLY$CDS,textposition = 'inside', name = 'CDS',
        hovertemplate = paste('(%{x}, %{y})'))
tracker <- ncol(CONT_TALLY) 
for(i in 4:tracker){
 FEAT_NAME <- colnames(CONT_TALLY)[i]
 CONT_COMP_FIG <- CONT_COMP_FIG %>% add_trace(x = CONT_TALLY[[i]],
                                              text = CONT_TALLY[[i]], textposition = 'inside', name = FEAT_NAME )
}
CONT_COMP_FIG <- CONT_COMP_FIG %>% layout(xaxis = list(title = "Gene count",
                                                        showgrid = TRUE,
                                                        showline = FALSE,
                                                        showticklabels = TRUE,
                                                        zeroline = FALSE,
                                                        domain = c(0.15, 1)),
                                                        yaxis = list(title = "Contigs",
                                                        categoryorder= 'total ascending',
                                                        showgrid = FALSE,
                                                        showline = FALSE,
                                                        showticklabels = TRUE,
                                                        zeroline = FALSE
                                                        ),
                                           barmode = 'stack',
                                           paper_bgcolor = 'rgb(255, 255, 255)', plot_bgcolor = 'rgb(255, 255, 255)',
                                           showlegend = TRUE) 
#### Length plot
CONT_LENGTH_FIG <- plot_ly(x = CONT_TALLY$cont_max, y = CONT_TALLY$X1, name = 'Contig length',width = 950, height = dim_high,
              type = 'bar', mode = 'markers',
              line = list(color = 'rgb(128, 0, 128)'),
              text=CONT_TALLY$cont_max, textposition = 'auto',
              hovertemplate = paste('(%{x}, %{y})')
              ) 
CONT_LENGTH_FIG <- CONT_LENGTH_FIG %>% layout(yaxis = list(
                      categoryorder= 'total ascending',
                      showgrid = TRUE, 
                      showline = TRUE, 
                      showticklabels = FALSE,
                      linecolor = 'rgb(102, 102, 102, 0.8)', 
                      domain = c(0.15, 1)),
                      xaxis = list(title = "Contig length",
                      zeroline = FALSE, 
                      showline = FALSE, 
                      showticklabels = TRUE, 
                      showgrid = TRUE,
                      side = 'down'),
         showlegend = FALSE) 
COMBINED_CONT_FIG <- subplot(CONT_COMP_FIG, CONT_LENGTH_FIG, shareY = TRUE, titleX = TRUE) 
COMBINED_CONT_FIG
```

#### Output table
Full BAKTA output table presented from GFF file with added "Name" column for gene names, extracted from "Attributes" column. For full description of columns please look at [link](https://en.wikipedia.org/wiki/General_feature_format)

Table 3 The table is fully interactive, searchable and regex compatible. It is possible to select shown columns by using "Column visiblity" tab. It is possible to download currently displayed table to a clipboard, csv, xlsx or pdf depending on user preferences.
```{r}
BAKTA_PROC <- subset(BAKTA_PROC, select = -c(X2,X6,X7,X8))
BAKTA_PROC <- BAKTA_PROC[, c(1,2,3,4,6,5)]
#remove contig description
BAKTA_PROC <- BAKTA_PROC %>% filter (X3 != "region")
DTABLE <- datatable(BAKTA_PROC, colnames = c("Contig",  "Feature", "Start", "End", "Name","Attributes"),  rownames = FALSE, class = 'hover', extensions  = 'Buttons',
                    options = list( dom = '<"top"fB>rt<"bottom"lp>i<"clear">', buttons = c('copy', 'csv', 'excel', 'pdf', 'colvis'), 
                      search = list(regex = TRUE, caseInsensitive = FALSE),
                      columnDefs = list(list(
                        targets = 5,
                        render = JS(
                          "function(data, type, row, meta) {",
                          "return type === 'display' && data.length > 6 ?",
                          "'<span title=\"' + data + '\">' + data.substr(0, 6) + '...</span>' : data;",
                          "}")
                      ))), callback = JS('table.page(3).draw(false);')
                    )
DTABLE
```
#### Links and Citation
Klappenbach JA, Dunbar JM, Schmidt TM. rRNA operon copy number reflects ecological strategies of bacteria. Appl Environ Microbiol. 2000;66(4):1328-1333. doi:10.1128/AEM.66.4.1328-1333.2000  

Citation for BAKTA can be found under [DOI: 10.1101/2021.09.02.458689](https://doi.org/10.1101/2021.09.02.458689).
