### Sourmash

Sourmash is a command-line tool to compare and classify DNA sequences based on "k-mers" ("DNA-words" of length "k").  
Classification was performed using the "Genome Taxonomy Database" (GTDB), provided by sourmash, containing ~ 258k reference genomes.

```{r, echo = FALSE, message = FALSE}
CLASSIFICATION_FILE <- "#RESULTSENV#"
METAGENOME_FILE <- "#RESULTSENVMETA#"
SAMPLE_NAME <- "#NAMEENV#"

library(dplyr)
library(ggplot2)
library(plotly)
library(stringr)

if (file.exists(CLASSIFICATION_FILE)) {
	CLASSIFICATION_DATA <-read.delim(CLASSIFICATION_FILE, sep=",", header=TRUE)
	CLASSIFICATION_DATA_EXIST <- TRUE
} else { CLASSIFICATION_DATA_EXIST <- FALSE }

if (CLASSIFICATION_DATA_EXIST) {
	SOURMASH_STATUS <- CLASSIFICATION_DATA[1,"status"]
	SOURMASH_SUPERKINGDOM <- strsplit(toString(CLASSIFICATION_DATA[1,"superkingdom"]), '__')[[1]][2]
	SOURMASH_PHYLUM <- strsplit(toString(CLASSIFICATION_DATA[1,"phylum"]), '__')[[1]][2]
	SOURMASH_CLASS <- strsplit(toString(CLASSIFICATION_DATA[1,"class"]), '__')[[1]][2]
	SOURMASH_ORDER <- strsplit(toString(CLASSIFICATION_DATA[1,"order"]), '__')[[1]][2]
	SOURMASH_FAMILY <- strsplit(toString(CLASSIFICATION_DATA[1,"family"]), '__')[[1]][2]
	SOURMASH_GENUS <- strsplit(toString(CLASSIFICATION_DATA[1,"genus"]), '__')[[1]][2]
	SOURMASH_SPECIES <- strsplit(toString(CLASSIFICATION_DATA[1,"species"]), '__')[[1]][2]
	SOURMASH_STRAIN <- strsplit(toString(CLASSIFICATION_DATA[1,"strain"]), '__')[[1]][2]
}

if (file.exists(METAGENOME_FILE)) {
	METAGENOMIC_DATA <-read.delim(METAGENOME_FILE, sep=",", header=TRUE)
	METAGENOMIC_DATA_EXIST <- TRUE
} else {
	METAGENOMIC_DATA_EXIST <- FALSE
	GREP_F_MATCH <- FALSE
}

if (METAGENOMIC_DATA_EXIST) {
	if (CLASSIFICATION_DATA_EXIST) {
		if (SOURMASH_SPECIES != "NA") {
    		F_MATCH_VALUE <- round(METAGENOMIC_DATA$f_match[grep(SOURMASH_SPECIES, METAGENOMIC_DATA$name)],2)
    		GREP_F_MATCH <- TRUE
    	} else if (SOURMASH_GENUS != "NA") {
    		F_MATCH_VALUE <- round(METAGENOMIC_DATA$f_match[grep(SOURMASH_GENUS, METAGENOMIC_DATA$name)],2)
    		GREP_F_MATCH <- TRUE
    	} else {
			METAGENOMIC_ORGANISM_WITH_HIGHEST_PROBABILITY <- str_remove(toString(strsplit(toString(METAGENOMIC_DATA$name[1]), ' ',)[[1]][2:3]), ',')
			F_MATCH_VALUE <- round(METAGENOMIC_DATA$f_match[1],2)
			GREP_F_MATCH <- FALSE
    	}
	} else {
		METAGENOMIC_ORGANISM_WITH_HIGHEST_PROBABILITY <- str_remove(toString(strsplit(toString(METAGENOMIC_DATA$name[1]), ' ',)[[1]][2:3]), ',')
		F_MATCH_VALUE <- round(METAGENOMIC_DATA$f_match[1],2)
		GREP_F_MATCH <- FALSE
  	}
}
```

`r if(CLASSIFICATION_DATA_EXIST) {
	sprintf("Status of the classification is: **%s**.", SOURMASH_STATUS)
	sprintf("The sample **%s** was identified as following:\n
**Table 1: Taxonomic classification of sample %s.**  NA means that sourmash was unable to classify this taxonomic rank.\n
Taxonomic rank|Identified taxonomy
-|-
Superkingdom|%s
Phylum|%s
Class|%s
Order|%s
Family|%s
Genus|%s
Species|%s
Strain|%s", 
	SAMPLE_NAME, SAMPLE_NAME, SOURMASH_SUPERKINGDOM, SOURMASH_PHYLUM, SOURMASH_CLASS, SOURMASH_ORDER,
	SOURMASH_FAMILY, SOURMASH_GENUS, SOURMASH_SPECIES, SOURMASH_STRAIN)
} else {
	sprintf("Classification of the sample **failed**.")
}`

`r if(METAGENOMIC_DATA_EXIST & GREP_F_MATCH) {
	sprintf("Probability for this organism is **%s** (Score is from 0 to 1; 1 means exact match).\n", F_MATCH_VALUE)
} else if (METAGENOMIC_DATA_EXIST & !GREP_F_MATCH) {
	sprintf("Organism with the highest probability based on the metagenomic analysis is: **%s** with a value of **%s** (Score is from 0 to 1; 1 means exact match).\n",
	METAGENOMIC_ORGANISM_WITH_HIGHEST_PROBABILITY, F_MATCH_VALUE)
}`

`r if (METAGENOMIC_DATA_EXIST) {
	sprintf("#### **Distribution of identified organisms.**\n\n  The following chart gives a overview about all identified 
	organisms in the sample. It is based on how many bases belong to each organism ( called \"intersect_bp\").  
	The diagram is interactive so that upon hovering over a part of it the organism and according intersecting bp are showed. Further 
	organisms can get masked by clicking their name in the legend. A download-button for saving the diagramm as \"png\"-file is 
	provided in the upper right corner while hovering over it.")
}`

```{r, echo = FALSE}
if (METAGENOMIC_DATA_EXIST) {
	ORGANISM_LIST <- list(METAGENOMIC_DATA$name)
	for (ELEMENT in 1:length(ORGANISM_LIST[[1]])) {
		ORGANISM_LIST[[1]][ELEMENT] <- str_remove(toString(strsplit(ORGANISM_LIST[[1]][ELEMENT], ' ')[[1]][2:3]), ',')
	}

	BP_COUNT_LIST <- list(METAGENOMIC_DATA$intersect_bp)
	COLUMN_NAME_LIST <- c("ORGANISM", "BP_COUNT")
	ORGANISM_BP_DATA <- data.frame(ORGANISM_LIST, as.numeric(unlist(BP_COUNT_LIST)))
	colnames(ORGANISM_BP_DATA) <- COLUMN_NAME_LIST

	CONTIG_PIE_CHART <- plot_ly(ORGANISM_BP_DATA, labels = ~ORGANISM, values = ~BP_COUNT, type = 'pie',
                		insidetextorientation = 'radial',
                		textposition = 'inside',
                		textinfo = 'percent',
                		insidetextfont = list(color = '#FFFFFF'),
                		hoverinfo = 'label+text',
                		text = ~paste(BP_COUNT, ' bp'),
                		marker = list(colors = colors,
                              		line = list(color = '#FFFFFF', width = 1))
					)
	CONTIG_PIE_CHART <- CONTIG_PIE_CHART %>% layout(
          				xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
          				yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE)
					)

	CONTIG_PIE_CHART
}
```

`r if (METAGENOMIC_DATA_EXIST) {
	sprintf("**Figure 1: Distribution of all identified organisms in the sample %s.**   
	    The chart is divided by the number of bases belonging to each identified organism. \n\n", SAMPLE_NAME)
}`

#### **Citation**

When publishing this data cite sourmash as follows:
"Brown et al, (2016), sourmash: a library for MinHash sketching of DNA,
Journal of Open Source Software, 1(5), 27, doi:10.21105/joss.00027"

The paper can be found under [DOI: 10.21105/joss.00027](https://www.doi.org/10.21105/joss.00027).
Code and further information can be found in the [github-repository](https://github.com/sourmash-bio/sourmash).