### Sourmash

Sourmash is a command-line tool to compare and classify DNA sequences based on "k-mers" ("DNA-words" of length "k").   
Classification was performed using the "Genome Taxonomy Database" (GTDB), provided by sourmash, containing ~ 258k reference genomes.  

Sourmash version used: #VERSIONENV#  
Results are stored in: #PATHENV#  

```{r, echo = FALSE, message = FALSE, error=TRUE}
CLASSIFICATION_FILE <- "#RESULTSENV#"
METAGENOME_FILE <- "#RESULTSENVMETA#"
SAMPLE_NAME <- "#NAMEENV#"

library(dplyr)
library(ggplot2)
library(plotly)
library(stringr)

CLASSIFICATION_DATA_EXIST <- FALSE
METAGENOMIC_DATA_EXIST <- FALSE
METAGENOMIC_DATA_NOT_EMPTY <- FALSE
GREP_F_MATCH <- FALSE

if (file.exists(CLASSIFICATION_FILE)) {
	if (!file.info(CLASSIFICATION_FILE)$size == 0) {
		CLASSIFICATION_DATA <-read.table(CLASSIFICATION_FILE, sep=",", header=TRUE, na.strings = c("", "NA"))
		CLASSIFICATION_DATA_EXIST <- TRUE
	}
} 

if (CLASSIFICATION_DATA_EXIST) {
	SOURMASH_STATUS <- CLASSIFICATION_DATA[1,"status"]
	SOURMASH_SUPERKINGDOM <- strsplit(toString(CLASSIFICATION_DATA[1,"superkingdom"]), '__')[[1]][2]
	SOURMASH_PHYLUM <- strsplit(toString(CLASSIFICATION_DATA[1,"phylum"]), '__')[[1]][2]
	SOURMASH_CLASS <- strsplit(toString(CLASSIFICATION_DATA[1,"class"]), '__')[[1]][2]
	SOURMASH_ORDER <- strsplit(toString(CLASSIFICATION_DATA[1,"order"]), '__')[[1]][2]
	SOURMASH_FAMILY <- strsplit(toString(CLASSIFICATION_DATA[1,"family"]), '__')[[1]][2]
	SOURMASH_GENUS <- strsplit(toString(CLASSIFICATION_DATA[1,"genus"]), '__')[[1]][2]
	SOURMASH_SPECIES <- strsplit(toString(CLASSIFICATION_DATA[1,"species"]), '__')[[1]][2]
	SOURMASH_STRAIN <- strsplit(toString(CLASSIFICATION_DATA[1,"strain"]), '__')[[1]][2]
}

if (file.exists(METAGENOME_FILE)) {
	if (!file.info(METAGENOMIC_DATA)$size == 0) {
		METAGENOMIC_DATA <- read.table(METAGENOME_FILE, sep=",", header=TRUE, na.strings = c("", "NA"))
		METAGENOMIC_DATA_EXIST <- TRUE
	}
} 

if (METAGENOMIC_DATA_EXIST) {
	if (!is.na(METAGENOMIC_DATA$f_unique_to_query[1]) & !is.na(METAGENOMIC_DATA$name[1])) {
		METAGENOMIC_DATA_NOT_EMPTY <- TRUE
		METAGENOMIC_ORGANISM_WITH_HIGHEST_PROBABILITY <- str_remove(toString(strsplit(toString(METAGENOMIC_DATA$name[1]), ' ',)[[1]][2:3]), ',')
		F_UNIQUE_TO_QUERY_ORGANISM_WITH_HIGHEST_PROBABILITY <- METAGENOMIC_DATA$f_unique_to_query[1]
		F_UNIQUE_TO_QUERY_PERCENTAGE <- round((F_UNIQUE_TO_QUERY_ORGANISM_WITH_HIGHEST_PROBABILITY * 100), 1)
	}
}
```

`r if(CLASSIFICATION_DATA_EXIST) {
	ifelse(SOURMASH_STATUS == "found", sprintf("Status of the classification is: <span style='color:green'>**%s**</span>.\n", SOURMASH_STATUS),
		ifelse(SOURMASH_STATUS == "disagree", sprintf("Status of the classification is: <span style='color:orange'>**%s**</span>.\n", SOURMASH_STATUS),
			sprintf("Status of the classification is: <span style='color:red'>**%s**</span>.\n", SOURMASH_STATUS)))
} else {
	sprintf("Classification of the sample **failed**.")
}`

`r if(CLASSIFICATION_DATA_EXIST) {
	sprintf("The sample **%s** was identified as following:\n
**Table 1: Taxonomic classification of sample %s.**  NA means that sourmash was unable to classify this taxonomic rank.\n
Taxonomic rank|Identified taxonomy
-|-
Superkingdom|%s
Phylum|%s
Class|%s
Order|%s
Family|%s
Genus|%s
Species|%s
Strain|%s", 
	SAMPLE_NAME, SAMPLE_NAME, SOURMASH_SUPERKINGDOM, SOURMASH_PHYLUM, SOURMASH_CLASS,
	SOURMASH_ORDER, SOURMASH_FAMILY, SOURMASH_GENUS, SOURMASH_SPECIES, SOURMASH_STRAIN)
}`

`r if (METAGENOMIC_DATA_EXIST) {
	if (METAGENOMIC_DATA_EXIST & METAGENOMIC_DATA_NOT_EMPTY) {
		sprintf("#### **Distribution of identified organisms.**\n\n  The following chart gives a overview about all identified 
		organisms in the sample. It is based on how many bases of the matched organism were found in the analysed sequence ( called \"f_unique_to_query\").  
		The diagram is interactive so that upon hovering over a part of it the organism and according the according percentage of unique bases in the 
		anaylsed sequence are showed. 
		Further organisms can get masked by clicking their name in the legend. A download-button for saving the diagramm as \"png\"-file is 
		provided in the upper right corner while hovering over it.")
	} else {
		sprintf("Metagenomic analysis failed. Probability could not be calculated.")
	}
}`

`r if (METAGENOMIC_DATA_EXIST) {
	if (METAGENOMIC_DATA_EXIST & METAGENOMIC_DATA_NOT_EMPTY) {
		ifelse(F_UNIQUE_TO_QUERY_PERCENTAGE < 10, sprintf("<span style='color:red'>Warning, less than 10 percent could be taxonomically assigned to the analysed genome:\nThe organism might not be a bacteria/archea, is completely unknown without any close relatives, or the sample contains lots of contamination.</span>"),
			ifelse(F_UNIQUE_TO_QUERY_PERCENTAGE < 50, sprintf("<span style='color:red'>Warning, less than 50 percent could be taxonomically assigned to analysed genome:\nThe organism might not be a bacteria/archea, is unknown, or the sample contains lots of contamination.</span>"),
				ifelse(F_UNIQUE_TO_QUERY_PERCENTAGE < 80, sprintf("<span style='color:red'>Warning, less than 80 percent could be taxonomically assigned to the analysed genome:\nThe organism might be an unknown bacteria/archea species or the sample contains to much contamination.</span>"),
					sprintf("The percentage of unique bases of %s in the analysed sequence amounts to %s percent.", METAGENOMIC_ORGANISM_WITH_HIGHEST_PROBABILITY, F_UNIQUE_TO_QUERY_PERCENTAGE))))
	}
}`

```{r, echo = FALSE}
if (METAGENOMIC_DATA_EXIST) {
	if (METAGENOMIC_DATA_NOT_EMPTY) {
		ORGANISM_LIST <- list(METAGENOMIC_DATA$name)
		for (ELEMENT in 1:length(ORGANISM_LIST[[1]])) {
			ORGANISM_LIST[[1]][ELEMENT] <- str_remove_all(toString(strsplit(ORGANISM_LIST[[1]][ELEMENT], ' ')[[1]][1:3]), ',')
		}

		F_UNIQUE_TO_QUERY_LIST <- lapply(list(METAGENOMIC_DATA$f_unique_to_query),"*",100)
		PERCENTAGE_UNKNOWN <- 100 - sum(unlist(F_UNIQUE_TO_QUERY_LIST))
		
		if (PERCENTAGE_UNKNOWN > 0) {
			ORGANISM_LIST <- unlist(c(ORGANISM_LIST, "Unknown/Not found"))
		}

		if (PERCENTAGE_UNKNOWN > 0) {
			F_UNIQUE_TO_QUERY_LIST <- c(F_UNIQUE_TO_QUERY_LIST, PERCENTAGE_UNKNOWN)
		}

		COLUMN_NAME_LIST <- c("ORGANISM", "PERCENTAGE_UNIQUE_BASES")
		ORGANISM_PERCENTAGE_DATA <- data.frame(ORGANISM_LIST, as.numeric(unlist(F_UNIQUE_TO_QUERY_LIST)))
		colnames(ORGANISM_PERCENTAGE_DATA) <- COLUMN_NAME_LIST

		COLOR_LIST <- vector()
		COLORS <- c("#636efa","#ef553b","#00cc96","#d62728","#9467bd","#8c564b","#e377c2","#7f7f7f","#bcbd22","#17becf")
		if ("Unknown/Not found" %in% ORGANISM_LIST) {
			for(ELEMENT in 1:(length(ORGANISM_LIST)-1)) {
  				if (ELEMENT <= 10) {
					COLOR_LIST[ELEMENT] <- COLORS[ELEMENT]
				} else {
					ELEMENT_LAST_NUMBER <- str_sub(ELEMENT,-1,-1)
					if (ELEMENT_LAST_NUMBER == 0) {
						ELEMENT_LAST_NUMBER <- 10
					}
					COLOR_LIST[ELEMENT] <- COLORS[ELEMENT_LAST_NUMBER]
				}
			}
			COLOR_LIST[(length(COLOR_LIST)+1)] <- "#404040"
		} else {
			for(ELEMENT in 1:(length(ORGANISM_LIST)-1)) {
				if (ELEMENT <= 10) {
					COLOR_LIST[ELEMENT] <- COLORS[ELEMENT]
				} else {
					ELEMENT_LAST_NUMBER <- str_sub(ELEMENT,-1,-1)
					if (ELEMENT_LAST_NUMBER == 0) {
						ELEMENT_LAST_NUMBER <- 10
					}
					COLOR_LIST[ELEMENT] <- COLORS[ELEMENT_LAST_NUMBER]
				}
			}
		}

		CONTIG_PIE_CHART <- plot_ly(ORGANISM_PERCENTAGE_DATA, labels = ~ORGANISM, values = ~PERCENTAGE_UNIQUE_BASES, type = 'pie',
	                		insidetextorientation = 'radial',
	                		textposition = 'inside',
	                		textinfo = 'percent',
	                		insidetextfont = list(color = '#FFFFFF'),
	                		hoverinfo = 'label+text',
	                		text = ~paste(PERCENTAGE_UNIQUE_BASES, ' %'),
	                		marker = list(colors = COLOR_LIST,
	                              		line = list(color = '#FFFFFF', width = 1))
						)
		CONTIG_PIE_CHART <- CONTIG_PIE_CHART %>% layout(
	          				xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
	          				yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
							legend = list(orientation = "h",   # show entries horizontally
										xanchor = "center",  # use center of legend as anchor
										x = 0.5)
							)

		CONTIG_PIE_CHART
	}
}
```

`r if (METAGENOMIC_DATA_EXIST) {
	if (METAGENOMIC_DATA_NOT_EMPTY) {
		sprintf("**Figure 1: Distribution of all identified organisms in the sample %s.**   
		    The chart is divided by the number of bases belonging to each identified organism. \n\n", SAMPLE_NAME)
	}
}`

#### **Citation**

When publishing this data cite sourmash as follows:
"Brown et al, (2016), sourmash: a library for MinHash sketching of DNA,
Journal of Open Source Software, 1(5), 27, doi:10.21105/joss.00027"

The paper can be found under [DOI: 10.21105/joss.00027](https://www.doi.org/10.21105/joss.00027).
Code and further information can be found in the [github-repository](https://github.com/sourmash-bio/sourmash).
