

```{r, echo = FALSE, message = FALSE}
library(dplyr)
library(ggplot2)
library(plotly)
library(stringr)

if (file.exists("/input/61405654fce5c90bc660e0db.consensus_taxonomy.tsv")) {
sourmash_data <-read.delim("/input/61405654fce5c90bc660e0db.consensus_taxonomy.tsv", sep=",", header=TRUE)
} else { message("Input-file from sourmash-classification process missing!") }

if (file.exists("/input/61405654fce5c90bc660e0db.consensus__composition.csv")) {
metagenomic_data <-read.delim("/input/61405654fce5c90bc660e0db.consensus__composition.csv", sep=",", header=TRUE)
} else { message("Input-file from sourmash-metagenome process missing!") }
```
Status of the classification is: **`r sourmash_data[1,"status"]`**.

The sample **#SAMPLE_IDENTIFICATION_NAME#** was identified as following:

Level|Classification
-|-
Superkingdom|`r strsplit(sourmash_data[1,"superkingdom"], '__')[[1]][2]`
Phylum|`r strsplit(sourmash_data[1,"phylum"], '__')[[1]][2]`
Class|`r strsplit(sourmash_data[1,"class"], '__')[[1]][2]`
Order|`r strsplit(sourmash_data[1,"order"], '__')[[1]][2]`
Family|`r strsplit(sourmash_data[1,"family"], '__')[[1]][2]`
Genus|`r strsplit(sourmash_data[1,"genus"], '__')[[1]][2]`
Species|`r strsplit(sourmash_data[1,"species"], '__')[[1]][2]`
Strain|`r strsplit(toString(sourmash_data[1,"strain"]), '__')[[1]][2]`

Probability for this organism is **`r metagenomic_data$f_match[grep(strsplit(sourmash_data[1,"species"], '__')[[1]][2], metagenomic_data$name)]`**.

## Contigs

Distribution of bp according to the found organisms:
```{r, echo = FALSE}
organism_list <- list(metagenomic_data$name)
for (element in 1:length(organism_list[[1]])) {
  organism_list[[1]][element] <- str_remove(toString(strsplit(organism_list[[1]][element], ' ')[[1]][2:3]), ',')
}

bp_count_list <- list(metagenomic_data$intersect_bp)
column_name_list <- c("organism", "bp_count")
organism_bp_data <- data.frame(organism_list, as.numeric(unlist(bp_count_list)))
colnames(organism_bp_data) <- column_name_list

pie_chart <- plot_ly(organism_bp_data, labels = ~organism, values = ~bp_count, type = 'pie',
              insidetextorientation = 'radial',
              textposition = 'inside',
              textinfo = 'label+percent',
              insidetextfont = list(color = '#FFFFFF'),
              hoverinfo = 'label+text',
              text = ~paste(bp_count, ' bp'),
              marker = list(colors = colors,
                            line = list(color = '#FFFFFF', width = 1)),
              showlegend = FALSE)
pie_chart <- pie_chart %>% layout(
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

pie_chart
```

## Command & Version

Sourmash-Version: #SOURMASH_VERSION#

Following commands were used to generate this results:

```
insert commands
```

## Citation

Citation for Sourmash can be found under [DOI: 10.21105/joss.00027](https://www.doi.org/10.21105/joss.00027).

